sdk/java/pulumi/src/main/java/com/pulumi/resources/StackReferenceOutputDetails.java

```diff
@@ -0,0 +1,92 @@
+package com.pulumi.resources;
+
+import java.util.Optional;
+
+/**
+ * StackReferenceOutputDetails records an output from a StackReference.
+ * <p>
+ * At most one of value and secretValue will be set.
+ */
+public class StackReferenceOutputDetails {
```

Comment:
Yeah, that's what we're doing here!
The constructor is protected so nobody can instantiate this class directly -- without using the Builder, which is public.
(We shouldn't make it completely impossible to build valid instances of this class because it's a plain data object—they should be able to build an object of this kind to return from mocks/use cases that we didn't consider.)

Reply:
To clarify: is it possible to make this class private and inside the StackReference class so users can't instantiate it directly using the methods that we're using.

Reply:
Sorry, first I need to clarify some terminology: in the context of Java, I'm interpreting 'private' as `private class Foo` or `private T method` — something that cannot be referenced outside that class. We cannot make this class private because it _must_ be possible to reference the type directly; otherwise a user cannot, say, write a function that returns the result of calling `myStackReference.outputDetailsAsync("foo")`, because the type is private.

We _can_ nest the class inside StackReference and make it `public static` so `StackReference.OutputDetails` instead of `StackReferenceOutputDetails` but honestly, that doesn't align with the patterns in the rest of the code and doesn't add much value besides making the class annoying to use (you'd have to type `StackReference.O<tab>` instead of `StackRef<tab>`, because you probably don't want to reference `OutputDetails` directly because that'll take up a much more generic name in the scope which—if you wanted that, why nest the class at all; sorry this is a bit of a rant about nested classes).

RE: "so users can't instantiate it". That's something we can but should not do. StackReferenceOutputDetails is a plain data object with no internal state or special behavior—plain struct equivalent from Go. We should not forbid users from instantiating one because we want users to be able to do "things" with it. Like, they may want to take a StackReferenceOutputDetails object that contains a `Map<String, Object>` and build one that contains a `MyBucketInfo` and pass that down the line. We also want users to be able to write unit tests for code that uses the outputDetailsAsync function for which being able to construct and return a StackReferenceOutputDetails from a mock StackReference would be valuable.

I agree that reducing surface area is desirable, but in this case, the machinery to build a plain data object is part of the definition of the data object. If we were using, say, Lombok, this would be `@Data @Builder` and no code.
(FWIW, other data objects also do not hide this: E.g. [CustomTimeouts](https://github.com/pulumi/pulumi-java/blob/main/sdk/java/pulumi/src/main/java/com/pulumi/resources/CustomTimeouts.java) (although it left its constructor public), [ResourceOptions](https://github.com/pulumi/pulumi-java/blob/main/sdk/java/pulumi/src/main/java/com/pulumi/resources/ResourceOptions.java), and [Alias](https://github.com/pulumi/pulumi-java/blob/main/sdk/java/pulumi/src/main/java/com/pulumi/core/Alias.java).)

Reply:
true. mocks are nice.